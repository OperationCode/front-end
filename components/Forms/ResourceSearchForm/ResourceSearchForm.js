import React from 'react';
import { Field, Formik, Form } from 'formik';
import omit from 'lodash/omit';
import styles from 'styles/resources.module.css';
import { string, func, shape, array } from 'prop-types';
import Button from 'components/Buttons/Button/Button';
import Input from 'components/Form/Input/Input';
import Select from 'components/Form/Select/Select';
import {
  RESOURCE_SEARCH,
  RESOURCE_SEARCH_BUTTON,
  RESOURCE_RESET_BUTTON,
} from 'common/constants/testIDs';

ResourceSearchForm.propTypes = {
  setIsLoading: func.isRequired,
  updateQuery: func.isRequired,
  setErrorMessage: func.isRequired,
  allLanguages: array.isRequired,
  allCategories: array.isRequired,
  fields: shape({
    languages: string,
    category: string,
    paid: string,
    q: string,
  }).isRequired,
};

function ResourceSearchForm({
  fields,
  setIsLoading,
  updateQuery,
  setErrorMessage,
  allCategories,
  allLanguages,
}) {
  const initialValues = {
    category: fields.category || '',
    q: fields.q || '',
    languages: Array.isArray(fields.languages)
      ? fields.languages
      : [fields.languages].filter(Boolean),
    free: fields.free || false,
  };

  const costOptions = [
    { value: 'false', label: 'Paid' },
    { value: 'true', label: 'Free' },
  ];

  const handleSubmit = (values, actions) => {
    setIsLoading(true);
    const emptyQueryParameters = Object.entries(values).filter(
      item => item[1] === null || !item[1].length,
    );
    const activeParameters = omit(
      values,
      emptyQueryParameters.map(parameter => parameter[0]),
    );

    updateQuery(activeParameters);
    setTimeout(() => {
      actions.setSubmitting(false);
    }, 500);
  };

  const handleReset = () => {
    setErrorMessage(null);
    updateQuery('');
  };

  return (
    <Formik
      enableReinitialize
      initialValues={initialValues}
      onSubmit={(values, actions) => {
        handleSubmit(values, actions);
        actions.setSubmitting(true);
      }}
    >
      {({ isSubmitting }) => (
        <Form role="search" onReset={handleReset}>
          <Field
            hasValidationStyling={false}
            data-testid={RESOURCE_SEARCH}
            disabled={isSubmitting}
            type="search"
            name="q"
            label="Search Keywords"
            component={Input}
          />

          <div className={styles.formContainer}>
            <div className={styles.selectColumn}>
              <Field
                hasValidationStyling={false}
                isDisabled={isSubmitting}
                placeholder="Start typing a category..."
                label="By Category"
                name="category"
                options={allCategories}
                component={Select}
              />
            </div>

            <div className={styles.selectColumn}>
              <Field
                hasValidationStyling={false}
                isDisabled={isSubmitting}
                placeholder="Resource cost..."
                label="By Cost"
                name="free"
                options={costOptions}
                component={Select}
              />
            </div>

            <div className={styles.selectColumn}>
              <Field
                hasValidationStyling={false}
                isDisabled={isSubmitting}
                placeholder="Start typing a language..."
                isMulti
                label="By Language(s)"
                name="languages"
                options={allLanguages}
                component={Select}
              />
            </div>
          </div>

          <div className={styles.buttonGroup}>
            <Button
              className={styles.buttonSingle}
              data-testid={RESOURCE_SEARCH_BUTTON}
              disabled={isSubmitting}
              theme="secondary"
              type="submit"
            >
              Search
            </Button>

            <Button
              className={styles.buttonSingle}
              data-testid={RESOURCE_RESET_BUTTON}
              disabled={isSubmitting}
              theme="secondary"
              type="reset"
            >
              Reset
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}

export default ResourceSearchForm;
