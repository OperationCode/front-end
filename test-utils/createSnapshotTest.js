/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
