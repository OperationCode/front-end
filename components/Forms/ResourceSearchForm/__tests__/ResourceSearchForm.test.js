import React from 'react';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import ResourceSearchForm from '../ResourceSearchForm';

const mockFields = {
  languages: null,
  category: null,
  paid: null,
  q: null,
};

describe('ResourceSearchForm', () => {
  it('should render with required props', () => {
    createSnapshotTest(
      <ResourceSearchForm
        fields={mockFields}
        setIsLoading={jest.fn()}
        updateQuery={jest.fn()}
        setErrorMessage={jest.fn()}
        allCategories={[]}
        allLanguages={[]}
      />,
    );
  });
});

// TODO Write more tests for this form
