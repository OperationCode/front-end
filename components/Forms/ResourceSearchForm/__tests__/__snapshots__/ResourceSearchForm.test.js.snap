// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceSearchForm > should render with required props 1`] = `
<form
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
  role="search"
>
  <div
    className="field"
    data-testid="INPUT"
  >
    <label
      className="text-themeSecondary text-sm"
      data-testid="LABEL"
      htmlFor="q"
    >
      Search Keywords
    </label>
    <div
      className="inputFeedbackGrouping"
      data-testid="INPUT_FEEDBACK_GROUPING"
    >
      <input
        className="Input"
        data-testid="RESOURCE_SEARCH"
        disabled={false}
        id="q"
        name="q"
        onBlur={[Function]}
        onChange={[Function]}
        type="search"
        value=""
      />
    </div>
  </div>
  <div
    className="formContainer"
  >
    <div
      className="selectColumn"
    >
      <div
        className="field"
      >
        <label
          className="text-themeSecondary text-sm"
          data-testid="LABEL"
          htmlFor="category"
        >
          By Category
        </label>
        <div
          className="selectFeedbackGrouping"
        >
          <div
            className=" css-b62m3t-container"
            onKeyDown={[Function]}
          >
            <span
              className="css-1f43avz-a11yText-A11yText"
              id="react-select-category-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              className="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              className=" css-8olv4-control"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <div
                className=" css-art2ul-ValueContainer2"
              >
                <div
                  className=" css-1jqq78o-placeholder"
                  id="react-select-category-placeholder"
                >
                  Start typing a category...
                </div>
                <div
                  className=" css-w9q2zk-Input2"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-category-placeholder"
                    aria-expanded={false}
                    aria-haspopup={true}
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    className=""
                    disabled={false}
                    id="react-select-category-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    role="combobox"
                    spellCheck="false"
                    style={
                      {
                        "background": 0,
                        "border": 0,
                        "color": "inherit",
                        "font": "inherit",
                        "gridArea": "1 / 2",
                        "label": "input",
                        "margin": 0,
                        "minWidth": "2px",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "100%",
                      }
                    }
                    tabIndex={0}
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className=" css-4xgw5l-IndicatorsContainer2"
              >
                <span
                  className=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  className=" css-itfj09-indicatorContainer"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="css-tj5bde-Svg"
                    focusable="false"
                    height={20}
                    viewBox="0 0 20 20"
                    width={20}
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="category"
              type="hidden"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="selectColumn"
    >
      <div
        className="field"
      >
        <label
          className="text-themeSecondary text-sm"
          data-testid="LABEL"
          htmlFor="free"
        >
          By Cost
        </label>
        <div
          className="selectFeedbackGrouping"
        >
          <div
            className=" css-b62m3t-container"
            onKeyDown={[Function]}
          >
            <span
              className="css-1f43avz-a11yText-A11yText"
              id="react-select-free-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              className="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              className=" css-8olv4-control"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <div
                className=" css-art2ul-ValueContainer2"
              >
                <div
                  className=" css-1jqq78o-placeholder"
                  id="react-select-free-placeholder"
                >
                  Resource cost...
                </div>
                <div
                  className=" css-w9q2zk-Input2"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-free-placeholder"
                    aria-expanded={false}
                    aria-haspopup={true}
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    className=""
                    disabled={false}
                    id="react-select-free-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    role="combobox"
                    spellCheck="false"
                    style={
                      {
                        "background": 0,
                        "border": 0,
                        "color": "inherit",
                        "font": "inherit",
                        "gridArea": "1 / 2",
                        "label": "input",
                        "margin": 0,
                        "minWidth": "2px",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "100%",
                      }
                    }
                    tabIndex={0}
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className=" css-4xgw5l-IndicatorsContainer2"
              >
                <span
                  className=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  className=" css-itfj09-indicatorContainer"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="css-tj5bde-Svg"
                    focusable="false"
                    height={20}
                    viewBox="0 0 20 20"
                    width={20}
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="free"
              type="hidden"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="selectColumn"
    >
      <div
        className="field"
      >
        <label
          className="text-themeSecondary text-sm"
          data-testid="LABEL"
          htmlFor="languages"
        >
          By Language(s)
        </label>
        <div
          className="selectFeedbackGrouping"
        >
          <div
            className=" css-b62m3t-container"
            onKeyDown={[Function]}
          >
            <span
              className="css-1f43avz-a11yText-A11yText"
              id="react-select-languages-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              className="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              className=" css-8olv4-control"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <div
                className=" css-art2ul-ValueContainer2"
              >
                <div
                  className=" css-1jqq78o-placeholder"
                  id="react-select-languages-placeholder"
                >
                  Start typing a language...
                </div>
                <div
                  className=" css-w9q2zk-Input2"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-languages-placeholder"
                    aria-expanded={false}
                    aria-haspopup={true}
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    className=""
                    disabled={false}
                    id="react-select-languages-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    role="combobox"
                    spellCheck="false"
                    style={
                      {
                        "background": 0,
                        "border": 0,
                        "color": "inherit",
                        "font": "inherit",
                        "gridArea": "1 / 2",
                        "label": "input",
                        "margin": 0,
                        "minWidth": "2px",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "100%",
                      }
                    }
                    tabIndex={0}
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className=" css-4xgw5l-IndicatorsContainer2"
              >
                <span
                  className=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  className=" css-itfj09-indicatorContainer"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="css-tj5bde-Svg"
                    focusable="false"
                    height={20}
                    viewBox="0 0 20 20"
                    width={20}
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div>
              <input
                name="languages"
                type="hidden"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="buttonGroup"
  >
    <button
      className="Button buttonSingle secondary"
      data-testid="RESOURCE_SEARCH_BUTTON"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="submit"
    >
      Search
    </button>
    <button
      className="Button buttonSingle secondary"
      data-testid="RESOURCE_RESET_BUTTON"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="reset"
    >
      Reset
    </button>
  </div>
</form>
`;
